### PR TITLE
PorousFlow: Enthalpy injection BC + Action

### PR DESCRIPTION
**Pull request by andrsd**
- [ ] Better name for the proxy BC use by the action
- [x] Jacobian test
- [ ] Regression test for `PorousFlowEnthalpySink`

### CODE REVIEW

**Comment by andrsd** [comment]
@rpodgorney @WilkAndy @cpgr This is the first draft for the enthalpy injection BC plus action that will easily set it up.

The BC is in the first patch, the action related code is in the second patch.

I need your guys' help with:

- better name for the BC (?)
- the test is dumb and should be improved, but I don't have enough knowledge to built it. I just put objects in it, so it would do the steps and made sure the BC objects are added.
- also, can this BC be used both for sink/source or just source? Do we need more checking somewhere?

There's probably more things you will spot right away. Let me know and I will fix those unless you want to throw additional patches on this branch which is perfectly fine with me.

**Comment by rpodgorney** [comment]
@WilkAndy @cpgr I'll let you weigh in on the naming, by @andrsd are we working on the BC or on the point source/sink. 

For a source, we need to specify the temperature, for a sink, we need to calculate the enthalpy being extracted from the domain

**Comment by WilkAndy** [comment]
I might get to this today, if not tomorrow.

**Comment by WilkAndy** [comment]
I think i like `PorousFlowEnthalpySink` better than what you've got, but it's not a deal breaker!  Maybe `PorousFlowHotFluidSink`?

File: modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C
Comment (WilkAndy) [inline modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C:35]: This is my main concern with this design.  Everywhere else we get "pressure" (and its derivatives) from a Material.  The reason for this is that "pressure" may not be a Variable.  Eg, when you're doing 2phase systems and you're using a (P,S) version, or Chris's crazy FluidState stuff that uses different variables.   See, eg, `PorousFlowSinkPTDefiner`

**Review by WilkAndy — CHANGES_REQUESTED** [review]
I might have more suggestions, but just wanted to quickly raise this "major" point.

**Comment by cpgr** [comment]
> I think i like `PorousFlowEnthalpySink` better than what you've got, but it's not a deal breaker! Maybe `PorousFlowHotFluidSink`?

I like `PorousFlowEnthalpySink`

**Comment by rpodgorney** [comment]
I’d go with PorousFlowEnthalpySink

Robert Podgorney, PhD, PG
Lead, Geothermal and Fossil Energy
Energy and Environment Division
Idaho National Laboratory

Office: 208-526-1524<tel:208-526-1524>
Cell: 208-520-9361<tel:208-520-9361>
email: robert.podgorney@inl.gov<mailto:robert.podgorney@inl.gov>

-sent from my mobile device, please excuse any typos or autocorrect errors

Statement of Confidentiality
The contents of this e-mail message and any attachments are confidential and are intended solely for addressee. The information may also be legally privileged. This transmission is sent in trust, for the sole purpose of delivery to the intended recipient. If you have received this transmission in error, any use, reproduction or dissemination of this transmission is strictly prohibited. If you are not the intended recipient, please immediately notify the sender by reply e-mail or phone and delete this message and its attachments, if any.


On May 26, 2020, at 18:50, Chris Green <notifications@github.com> wrote:

﻿

I think i like PorousFlowEnthalpySink better than what you've got, but it's not a deal breaker! Maybe PorousFlowHotFluidSink?

I like PorousFlowEnthalpySink

**Comment by andrsd** [comment]
I am assuming you guys are talking about renaming `PorousFlowSinkEnthalpy` to `PorousFlowEnthalpySink`.

What about the name for the bundled BC. Currently, the syntax looks like:
```
[PorousFlow]
  [./BCs]
    [./left]
      type = PorousFlowSourceBC
      boundary = left
      pressure = pp
      T_in = 300
      fp = simple_fluid
      flux_function = -1
    [../]
  [../]
[]
```

Is `PorousFlowSourceBC` ok? May be some more descriptive name would be better...

File: modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C
Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C:35]: Ok. If I get pressure from the material property, how do I get the jacobian? I think you guys compute the derivatives in the material, but that means I need to "replace" your version of the material on the side with my side material, which may not be even possible. Or do other convoluted things to work around this problem. I think that's why I did it via coupled variable, because it was way easier.

Also, I don't think @rpodgorney plans to use this with other than (p, T) formulation. But, if we want to generalize it, I would need your help with  - I don't understand all of the porous flow stuff. So, if you have suggestions how I should code this up, let me know. I am happy do to the work...

Comment (WilkAndy) [reply modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C]: We also calculate the derivatives you need:
```
 _dpp_dvar(getMaterialProperty<std::vector<std::vector<Real>>>("dPorousFlow_porepressure_nodal_dvar")
```
then use the chain rule with your dh_dp, eg
```
jac = dh_dp * dpp_dvar[i][ph][pvar]
```
(with phi and test too), where
```
 const unsigned int pvar = _dictator.porousFlowVariableNum(jvar);
```

Comment (WilkAndy) [reply modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C]: and i is the node and ph is the phase (user input)

Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C]: so if `jvar` corresponds to the temperature variable, is `dpp_dvar[i][ph][pvar] == 0`? (assuming `pvar = _dictator.porousFlowVariableNum(jvar);`)

Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C]: and then, I am assuming `dpp_dvar` is going to be 1 for my case with single pressure...

Comment (WilkAndy) [reply modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C]: > so if `jvar` corresponds to the temperature variable, is `dpp_dvar[i][ph][pvar] == 0`? (assuming `pvar = _dictator.porousFlowVariableNum(jvar);`)

I *think* this is true for all cases currently, yes.  I have a vague feeling that we were playing with entropy being a variable, and maybe people are even using that, in which case the derivative would be nonzero.

Comment (WilkAndy) [reply modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C]: > and then, I am assuming `dpp_dvar` is going to be 1 for my case with single pressure...

Yes

**Comment by andrsd** [comment]
`PorousFlowEnthalpySink` now takes pore pressure from the material... Relative error on jacobian test is ~1e-6, so I think the jacobian is implemented correctly, but double check it, please.

**Comment by WilkAndy** [comment]
> `PorousFlowEnthalpySink` now takes pore pressure from the material... Relative error on jacobian test is ~1e-6, so I think the jacobian is implemented correctly, but double check it, please.

I didn't get to this today, sorry, but i can't see any mistakes just glancing over it.  Can you wait for another <24 hours for me to do a more thorough review?

a

**Comment by andrsd** [comment]
> Can you wait for another <24 hours for me to do a more thorough review?

Yeah, no problem...

File: modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C
Comment (WilkAndy) [inline modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C:35]: This is a mass flux, so you need to specify its SI unit is kg.m^-2.s^-1.

File: modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C
Comment (WilkAndy) [inline modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C:78]: You mentioned that you checked this is correct.  Please include a test for it.

File: modules/porous_flow/include/actions/PorousFlowAddBCAction.h
Comment (WilkAndy) [inline modules/porous_flow/include/actions/PorousFlowAddBCAction.h:15]: Did you intend that this Action should be used for other BCs too?  If not, then change to "Action that adds BCs corresponding to hot/cold fluid injection, viz a PorousFlowSink and a PorousFlowEnthalpySink"

File: modules/porous_flow/include/actions/PorousFlowAddBCAction.h
Comment (WilkAndy) [inline modules/porous_flow/include/actions/PorousFlowAddBCAction.h:27]: Doxygen comment needed

File: modules/porous_flow/include/bcs/PorousFlowEnthalpySink.h
Comment (WilkAndy) [inline modules/porous_flow/include/bcs/PorousFlowEnthalpySink.h:19]: "Applies a flux sink of heat energy to a boundary..."

File: modules/porous_flow/include/bcs/PorousFlowEnthalpySink.h
Comment (WilkAndy) [inline modules/porous_flow/include/bcs/PorousFlowEnthalpySink.h:42]: "The mass flux"

File: modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C
Comment (WilkAndy) [inline modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C:61]: Please think about this: i believe you should have `_pp[_i][_ph]` because we probably want to use the upwinded version.  In your case i don't think it matters hugely since this is a source and not a sink (upwinding is especially important in sinks, because quadpoint stuff can inadvertently pull out too much fluid/heat resulting in negative fluid mass or temperature).  Also, you don't need to upwind any mobilities, etc.  But everywhere else we've used full upwinding, so i suggest to use it here.  If you want to use `_pp[_qp][_ph]` then you should pull in "PorousFlow_porepressure_qp" instead of "PorousFlow_porepressure_nodal" (and same for the derivative).

Then your code will be like line238 of `PorousFlowSink.C`.

File: modules/porous_flow/test/tests/sinks/s11.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/sinks/s11.i:95]: Shouldn't need this: it should get added automatically

File: modules/porous_flow/test/tests/sinks/s11.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/sinks/s11.i:72]: Can use PorousFlow1PhaseFullySaturated instead - i think i mentioned this before

File: modules/porous_flow/test/tests/sinks/s11.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/sinks/s11.i:1]: Probably don't need such a giant mesh for this case

File: modules/porous_flow/test/tests/sinks/s11.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/sinks/s11.i:30]: Can use PorousFlowCapillyPressureConst instead

File: modules/porous_flow/test/tests/sinks/s11.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/sinks/s11.i:135]: OK, i just commented on this test, but now i just realised you're probably going to tidy this up anyway, and add more tests (and include in the "tests" file) and doco.  Oh well....

File: modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C
Comment (WilkAndy) [inline modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C:88]: Again, think about upwinding.  If you use the upwinded version, these lines should be
```
if (_i != _j)
  return 0.0;
_fp.h_from_p_T(_pp[_i][_ph], _T_in, h, dh_dpp, dh_dT);
const Real jac = dh_dpp * _dpp_dvar[_i][_ph][pvar];
return _test[_i][_qp] * jac * _m_func.value(_t, _q_point[_qp])
```
If you don't want to upwind, then you should use `_dpp_dvar[_qp][_ph][pvar]` and include the `_m_func.value(_t, _q_point[_qp])`  (and pull in the "_qp" version of the derivative, as mentioned above)

**Review by WilkAndy — CHANGES_REQUESTED** [review]
Thanks @andrsd .  Lots of small comments.

File: modules/porous_flow/src/bcs/PorousFlowSourceBC.C
Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowSourceBC.C:28]: > this class is confusingly named,

Yes, that's just my ignorance, I will rename it to `PorousFlowSinkBC`, so it is consistent with the rest of the module. It is just a proxy for setting up `PorousFlowSink` and `PorousFlowEnthalpySink`, so having `Sink` in its name is appropriate.

File: modules/porous_flow/src/bcs/PorousFlowSourceBC.C
Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowSourceBC.C:40]: Yeah, I am really bad at writing doco...

File: modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C
Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C:39]: Nice 👍 . I would not be able to come up with that.

File: modules/porous_flow/src/bcs/PorousFlowSourceBC.C
Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowSourceBC.C:28]: Actually, I need a better name so it is not confusing why there is `PorousFlowSinkBC` and `PorousFlowSink`... 🤔

File: modules/porous_flow/include/actions/PorousFlowAddBCAction.h
Comment (andrsd) [reply modules/porous_flow/include/actions/PorousFlowAddBCAction.h:15]: > Did you intend that this Action should be used for other BCs too?

Yes. That's my plan. If there are other BCs like this one, we just add it here, so it is all nicely in one place :-)

File: modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C
Comment (andrsd) [reply modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C:61]: I think you are right. My original version was using just qp-based values and this did not get updated when I pulled in `PorousFlow_porepressure_nodal`. Still learning your code ;-)

File: modules/porous_flow/test/tests/sinks/s11.i
Comment (andrsd) [reply modules/porous_flow/test/tests/sinks/s11.i:135]: yeah, still need some help with tests. i can do the jacobian one with no problems i think. but, i would probably need to redesign this one so it actually checks that the BC works and fails if somebody messes with the code. if you have ideas, i am all one ear...

**Comment by andrsd** [comment]
Jacobian test added.

**Comment by andrsd** [comment]
And, hopefully, I addressed all of @WilkAndy's  comments.

Still thinking about the regression tests if it is good enough. Thoughts?

File: modules/porous_flow/test/tests/jacobian/tests
Comment (WilkAndy) [inline modules/porous_flow/test/tests/jacobian/tests:1081]: Kernels -> BCs

You cheat!  I didn't know that you could just write "issues = '#'" to fool CIVET.  I'll do that from now on, thanks!  <joke>

File: modules/porous_flow/test/tests/jacobian/esbc01.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/jacobian/esbc01.i:78]: This looks like an OK simple test.  It's sometimes quite tricky to devise tests that actually test your code (and not derivatives of enthalpy or whatever).  I assume you've set the test tolerance tight enough to be meaningful.  It would be good if 'flux_function' and/or pp varied across the element, and it would be good if this wasn't a square element, so that you can more convincingly demonstrate you're getting the right jacobian for the right node.

File: modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C
Comment (WilkAndy) [inline modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C:96]: I'm pretty sure you don't need a "phi" here.

**Review by WilkAndy — CHANGES_REQUESTED** [review]

**Comment by WilkAndy** [comment]
> And, hopefully, I addressed all of @WilkAndy's comments.
> 
> Still thinking about the regression tests if it is good enough. Thoughts?

Tests: i would add two tests (but do what you think is best)

 - First test does not use the Action, and pumps fluid into the BC with spatially-dependent flux_flunction and pp, and explicitly checks that the mass and energy increases correctly at each node (yea - a bit of calculation by hand is needed).

 - Second test does the same with the Action and demonstrates should result in identical behaviour.

a

**Pull request closed without merging**
